Improve usability of Select component

thangdinh opened the pull request:
Basically, I would like to make select component now tabbable.  I also changed the behavior of pressing any key will show the dropdown, because it seems to be uncommon to show the dropdown when we press special keys such as: shift, capslock, F1,F2...
Please help me review it @igillis @azirbel @hanakimis 
Thank you.

hanakimis commented on src/css/select.less:
```diff
@@ -23,6 +24,13 @@
   outline: none;
+
+  &:focus {
+    .ember-select-choice {
```
can we put these on the same line, so it reads: "&:focus .ember-select-choice"? It makes it easier to read.

Thanks!

hanakimis commented:
CSS LGTM, besides comment : )

azirbel commented on src/select.coffee:
```diff
@@ -341,7 +342,15 @@ Ember.AddeparMixins.ResizeHandlerMixin,
+    # and the keycode should be one of special keys or common characters
+    # 'A'..'Z', 'a..z','0..9'
+    if event.keyCode in [@KEY_CODES.ENTER, @KEY_CODES.SPACEBAR] or
+    event.keyCode in [@KEY_CODES.DOWN, @KEY_CODES.UP] or
```
indentation is weird here; these should be indented twice, right?

bjcohen replied:
@azirbel Do we have a style recommendation for something like this? Indenting subsequent lines of the condition means that the body of the conditional needs to be indented to 4 spaces, which is also weird.

addepar-andy replied:
for this specific case, i'd make a static map with all the keys that match, and then check that map.

for general cases, when you have 40 different clauses in your `if` statement, its better to put each clause into a variable name that makes sense and then use those names, so that the statement is more readable.

azirbel commented on src/select.coffee:
```diff
@@ -349,7 +358,14 @@ Ember.AddeparMixins.ResizeHandlerMixin,
+      event.preventDefault()
+
+  tabPressed: (event) ->
+    if @get('showDropdown') is yes
```
`if @get('showDropdown')`

azirbel replied:
Or, more concisely: `@send 'hideDropdown' if @get('showDropdown')`

azirbel replied:
you can do this in many places in the code.

thangdinh replied:
Done. Found as much as I can.

azirbel commented:
Thanks for this @thangdinh 

Tested it out and I think it's a clear improvement over the current behavior, but a few scenarios are still a bit weird:

* In my opinion, hitting ESC should unfocus the component
* In the multi-select component, pressing ENTER seems to remove focus from the component, so you can't type more
* Actually, the multi-select component seems pretty broken, I've gotten it into a state where it doesn't filter when I type ... looks like this happens when I select one option, then click back into the box to type another. This might not be your fault.

I don't understand what you mean by "it seems to be uncommon to show the dropdown when we press special keys such as: shift, capslock..." For me it looks like the dropdown is always shown when the component comes in focus, even before characters are typed. Why do we need the special cases for common characters?

thangdinh commented:
@azirbel Thanks Alex for your detailed comments. Here are my thoughts:
+ ESC (on select) should close the dropdown, shouldn't unfocus the component, on forms there should be at least one component gets focused. On popular websites such as gmail, ebay... it appears that way.
+ I tested it on multi select before I left on the other branch. But i haven't tested it this time. It's my bad. I'll do it.
About the dropdown's behavior when the select gets focused, similar to my first point, on popular websites the dropdown only appears when you hit some special keys such as ENTER, SPACEBAR, arrow keys or hit the first letter of one option inside the list. Apparently, I think it's a good behavior. We shouldn't show dropdown when we focus on it because it looks interrupting (UI/UX). Imaging that when you already selected one option, and you navigate back and forth (tab, shift tab) it keeps popping out for no good reason. Anyway, UI/UX is a very debatable topic so I guess our product/designing team will decide on how it behaves.
I don't have much experience with multi-select component, so I'm not sure if it follows the same behavior as above or not. Do you have any websites in mind having such components?

azirbel commented:
* Ok, I'm ok with that ESC behavior. But when I run your branch and use the select or multi-select, I seem to lose focus and the dropdown stays open - are you sure it's working the way you described?
* The second behavior also makes sense to me. So you're saying if I TAB to a select component, it should focus but not open the dropdown. Then when a key is pressed, it should open the dropdown. I think this behavior sounds good, but again when running your branch I'm seeing the dropdown open as soon as I TAB to the select component, so it doesn't seem to be working for me.

thangdinh commented:
@azirbel:  I'm checking on this. And fixing the multi-select component problem. I tested again the cases you mentioned, and it runs fine on select component. I am not sure why it didn't work on your computer.

thangdinh commented:
Could anyone help me review this PR again? Thanks.
@hanakimis, @azirbel, @igillis, @Dylnuge

hanakimis commented on dist/css/ember-widgets.css:
```diff
@@ -45,6 +45,14 @@
   display: inline-block;
   outline: none;
 }
+.ember-select:focus .ember-select-choice,
```
Are you sure you're supposed to edit the css files in dist, rather than src?

igillis replied:
this is the compiled version of the changes in `select.less`, and he does need to check this in

azirbel commented on src/css/select.less:
```diff
@@ -6,6 +6,8 @@
 -------------------------------------------------- */
 // Note: Important that this is synchronized with the list-view rowHeight
 @rowHeight: 26px;
+@input-border-focus: #4d90fe;
```
can we use variables here? If not, comment on where the values came from?

azirbel replied:
wow, I completely misread this. deleting my comment.

azirbel commented on src/multi_select.coffee:
```diff
@@ -47,10 +47,27 @@ Ember.Widgets.MultiSelectComponent = Ember.Widgets.SelectComponent.extend
     @get('placeholder') or @get('persistentPlaceholder')
   .property 'query', 'placeholder', 'persistentPlaceholder', 'selections.length'
 
+  # add css class to multiselect to highlight border
```
no need for this comment or the one on line 54, since your function names are good.

azirbel replied:
that said, `addBorderToMultiSelectContainer` and `removeBorderFromMultiSelectContainer` would be better

azirbel commented on src/multi_select.coffee:
```diff
@@ -47,10 +47,27 @@ Ember.Widgets.MultiSelectComponent = Ember.Widgets.SelectComponent.extend
+      if @get('parentView.showDropdown')
+        Ember.run.schedule 'afterRender', this, ->
+          @$().focus() if (@get('_state') or @get('state')) is 'inDOM'
+    , 'parentView.showDropdown'
```
add newline between this and `placeholder`

azirbel commented on src/multi_select.coffee:
```diff
@@ -98,12 +115,30 @@ Ember.Widgets.MultiSelectComponent = Ember.Widgets.SelectComponent.extend
   deletePressed: (event) ->
-    if event.target.selectionStart == 0
+    # check if the cursor is at the beginning and no text is selected
+    if event.target.selectionStart == 0 and event.target.selectionEnd == 0
```
not your fault since this code existed before, but use `is` instead of `==`

azirbel commented on src/multi_select.coffee:
```diff
@@ -98,12 +115,30 @@ Ember.Widgets.MultiSelectComponent = Ember.Widgets.SelectComponent.extend
+
+    # dynamically add css style to highlight the border if any elements inside
+    # the multiselect is focused
+    if @$()[0].contains(activeElement) or @$()[0]==activeElement
```
use `is`

azirbel replied:
is this code necessary? I tried deleting it and the border still seemed to appear/disappear as expected. If it's necessary please add a comment explaining the case that breaks without it

azirbel commented on src/select.coffee:
```diff
@@ -107,6 +107,21 @@ Ember.AddeparMixins.ResizeHandlerMixin,
+
+  # set up a map for accepted keys from multiple ranges of them
+  # these are keys to show dropdown when being pressed
+  createAcceptedKeySet: ->
```
why not make this a computed property? `acceptedKeys: Ember.computed -> ... .property()`

azirbel commented on src/select.coffee:
```diff
@@ -340,8 +360,13 @@ Ember.AddeparMixins.ResizeHandlerMixin,
+    # [Spacebar, Enter, Up, Down, 'A'..'Z','a..z','0..9']
+    acceptedKeys = @get 'acceptedKeys'
+    if acceptedKeys[event.keyCode]
+      return @set('showDropdown', yes) if not @get 'showDropdown'
```
nitpick, but `if not` -> `unless`

azirbel commented on src/select.coffee:
```diff
@@ -64,10 +64,10 @@ Ember.AddeparMixins.ResizeHandlerMixin,
 
-  tabindex: -1
+  # we need to set tabindex so that div responds to key events
+  tabindex: 0
```
this change is good, but you should add `tabindex: -1` to the MultiSelectComponent, because right now you have to press tab twice (once for the component and once for the text field inside). I just tested with `tabindex: -1` and it seems to work as expected.

azirbel replied:
here's a comment from a while ago which still sounds valid

cyril-sf replied:
@thangdinh This comment doesn't seem to have been addressed.

thangdinh replied:
@cyril-sf it's already addressed in MultiSelectComponent as suggested.

azirbel commented:
Not sure why, but "hitting esc to close the dropdown and re-select the component" is not working for me. When I hit esc, I lose focus and the dropdown stays open

azirbel commented:
LGTMBC, a few comments are important so let me know if you have questions

azirbel commented:
Actually just noticed something else, the cursor is a bit weird once you select one element in multi-select. So you select one thing and focus goes back to the component, at which point you CAN type, but it doesn't *look* like you can type because there is no cursor (focus is on the component, not the text field).

thangdinh commented:
Thanks @azirbel. I addressed all of the comments you mentioned. However, I could not reproduce this issue: "When I hit esc, I lose focus and the dropdown stays open". When you have time, would you please help me double check it? Steps to reproduce it will be appreciated. Thanks.

azirbel commented:
Sure. (1) Click inside multi select component; (2) type "af". Dropdown should appear. (3) hit ESC. focus is lost and the dropdown stays open.

Just got the above with the latest code from your branch.

On the other hand, the current state is better than what's on master (we could consider this an existing bug). If the fix is out of scope or too hard, you can skip it.

By the way, the tab behavior is really awesome. Thanks for working on it.

thangdinh commented:
Thanks @azirbel. I still cannot reproduce the ESC behavior. I don't know why. 
Could you help me double check this behavior when you have time, @igillis? Thanks.

igillis commented:
I just checked out this branch and saw the following in both Chrome 41.0.2272.76 (64-bit) and FF 35.0.1 on Mac OS X 10.9.5.

1) Click inside the multi select
2) Type 'af'. Dropdown appears.
3) Hit ESC.
4) Dropdown closes, focus remains on the button, 'af' still present in the textbox

Similarly for the normal select, except when you re-open the dropdown 'af' isn't present. @azirbel I wonder if you have a different version of something? Would be curious where it's failing for you -- do you reach `escapePressed`?

igillis commented:
@azirbel: I get your behavior unless I disable my vimium plugin...any chance you're running that?

azirbel commented:
........you are a debugging god

thangdinh commented:
Thanks @igillis. You are awesome!

thangdinh commented:
I just added keyboard interaction tests. Would you please help me review it? Thanks.
@igillis @cyril-sf @azirbel

igillis commented:
@thangdinh and I spoke offline -- can you comment again when you get rid of the unnecessary `wait`s

thangdinh commented:
@igillis done

igillis commented:
I'm getting
```
FAILED  - Test keyboard interaction	Died on test #2     at /Users/ian.gillis/ember-widgets2/tests/unit/select.js:59:1: undefined is not a function
	TypeError: undefined is not a function
	    at /Users/ian.gillis/ember-widgets2/tests/unit/select.js:70:36
```
on my machine -- does it work on yours?

thangdinh commented:
My bad. Could you please try it again?

On Thu, Apr 2, 2015 at 3:48 PM, igillis <notifications@github.com> wrote:

> I'm getting
>
> FAILED  - Test keyboard interaction Died on test #2     at /Users/ian.gillis/ember-widgets2/tests/unit/select.js:59:1: undefined is not a function
>     TypeError: undefined is not a function
>         at /Users/ian.gillis/ember-widgets2/tests/unit/select.js:70:36
>
> on my machine -- does it work on yours?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Addepar/ember-widgets/pull/166#issuecomment-89067843>.
>

igillis commented on src/multi_select.coffee:
```diff
@@ -98,12 +107,26 @@ Ember.Widgets.MultiSelectComponent = Ember.Widgets.SelectComponent.extend
+  keyDown: (event) ->
+    @_super(event)
+    activeElement = $(document.activeElement)[0]
+    if @$()[0].contains(activeElement) or @$()[0]==activeElement
```
nit: use `is` instead of == and use proper spacing

nit: local variable with meaningful name for `@$()[0]`

igillis commented on src/select.coffee:
```diff
@@ -195,6 +214,9 @@ Ember.AddeparMixins.ResizeHandlerMixin,
       if @get('parentView.showDropdown')
         Ember.run.schedule 'afterRender', this, ->
           @$().focus() if (@get('_state') or @get('state')) is 'inDOM'
+      # clear the query string when dropdown is hidden
```
:+1:

igillis commented:
Nice usability win @thangdinh. Thanks for your patience and for adding some much-needed tests.

LGTM besides comments (you can merge after fixing my comments)

cyril-sf commented on src/multi_select.coffee:
```diff
@@ -98,12 +107,27 @@ Ember.Widgets.MultiSelectComponent = Ember.Widgets.SelectComponent.extend
     @get('selections').removeObject item
 
+  focusIn: (event) ->
+    @addBorderMultiSelectContainer()
```
I would probably set a boolean ( `hasFocus` ) to `yes` / `no` and use `bind-attr` in the template to toggle the class

cyril-sf replied:
```hbs
<div {{bind-attr class="hasFocus:ember-select-multi-focus"}}>
```

thangdinh replied:
:+1: Thank you.

cyril-sf commented on src/select.coffee:
```diff
@@ -195,6 +214,9 @@ Ember.AddeparMixins.ResizeHandlerMixin,
           @$().focus() if (@get('_state') or @get('state')) is 'inDOM'
+      # clear the query string when dropdown is hidden
+      else
+        @set 'parentView.query', ''
```
it seems like you can access the query through the `value` property

thangdinh replied:
Do you mean doing this instead? 
```
@set 'value', ''
```

cyril-sf commented on src/select.coffee:
```diff
@@ -340,13 +367,18 @@ Ember.AddeparMixins.ResizeHandlerMixin,
+    # [Spacebar, Enter, Up, Down, 'A'..'Z','a..z','0..9']
+    acceptedKeys = @get 'acceptedKeys'
+    if acceptedKeys[event.keyCode]
+      return @set('showDropdown', yes) unless @get 'showDropdown'
```
why not test both conditions in the `if` statement?

cyril-sf commented on src/select.coffee:
```diff
@@ -354,14 +386,23 @@ Ember.AddeparMixins.ResizeHandlerMixin,
-    @send 'hideDropdown'
+    if @get('showDropdown')
+      @send 'hideDropdown'
+      @$()?.focus()
```
how can `@()` be null?

thangdinh replied:
It is somehow null in the scenario where we put it on a carousel popover, and after hitting Enter, the carousel select disappears, and @$() is null. So I just wanted to make sure it not happening for ESC too

cyril-sf replied:
I see.

It's probably cleaner to check the state of the component ( `if @get('isDestroyed') or @get('isDestroying')` ) at the beginning of the event handler

cyril-sf commented on tests/unit/multi_select.coffee:
```diff
@@ -0,0 +1,67 @@
+    .then ->
+      ok isVisible(find '.ember-select-results', multiSelectComponent, 'Dropdown list should appear')
+
+    .then ->
```
it should not be needed

cyril-sf commented on tests/unit/multi_select.coffee:
```diff
@@ -0,0 +1,67 @@
+      resultItems = find '.ember-select-search-choice', multiSelectComponent
+      currentText = $(resultItems[resultItems.length - 1]).text().trim()
+      # trim the special character "x" at the end of the select choice text
+      equal(currentText.substring(0, currentText.length - 2), selectedText)
```
Adding a description to the assertion helps to understand the test without reading the actual code

cyril-sf commented on tests/unit/multi_select.coffee:
```diff
@@ -0,0 +1,67 @@
+    .then ->
+      ok isVisible(find '.ember-select-results', multiSelectComponent, 'Dropdown list should appear')
+
+      keyEvent(multiSelectComponent, 'keydown', 27)
```
which key is 27?

cyril-sf commented:
This is looking good. I'm happy to see some tests added :smile: 

Let me know if you have some questions about my comments.

thangdinh commented:
could anyone help me review one more round? @cyril-sf @azirbel @igillis 
Thanks.

azirbel commented on src/mixins.coffee:
```diff
@@ -1,3 +1,16 @@
+Ember.Widgets.KeyboardHelper = Ember.Mixin.create
+  KEY_CODES:
+    BACKSPACE: 8,
```
you could delete commas here since it's CS. nitpick.

thangdinh replied:
done

azirbel commented on src/multi_select.coffee:
```diff
@@ -101,12 +114,33 @@ Ember.Widgets.MultiSelectComponent = Ember.Widgets.SelectComponent.extend
 
   removeSelectItem: (item) ->
+    # set the focus back to the searchView because this item will be removed
+    currentDropDownState = @get('showDropdown')
```
another nitpick, but this is a weird name. Something like `dropdownIsShowing` would make line 125 read better.

thangdinh replied:
done

azirbel commented on src/multi_select.coffee:
```diff
@@ -101,12 +114,33 @@ Ember.Widgets.MultiSelectComponent = Ember.Widgets.SelectComponent.extend
+  enterPressed: (event) ->
+    @_super(event)
+    @focusTextField()
+    @set('showDropdown', yes) unless @get('showDropdown')
```
if it's already `yes`, there should be no harm setting it to `yes`, right? Can we remove the `unless`?

azirbel replied:
same below in `click`. I wonder if I'm missing something?

thangdinh replied:
I think it comes from `legacy` code when I was new to ember and coffee script. I saw one place having it and i was afraid of observers and such, that's why I put the check in to make sure :-)

azirbel replied:
ok, it's not a big deal but if you are able to try it out with the simpler versions of code that would be awesome :+1:

azirbel commented on src/select.coffee:
```diff
@@ -44,7 +44,12 @@ Ember.Widgets.SelectOptionView = Ember.ListItemView.extend
     if @get('controller.showDropdown')
-      @get('controller').send 'hideDropdown'
+      # Keep showing dropdown if it is multi select component
+      unless @get('controller.layoutName') is 'multi-select'
```
this is pretty bad, we should definitely find a way to override `click` in `multi-select` so that `select` doesn't have to have logic for `multi-select`.

I'd recommend refactoring lines 40-42 into a `selectContent` function, calling that in the `click` handler for both `select` and `multi-select`, and then having each component handle its own logic for hiding the dropdown. You could then remove the `@_super` call from the `multi-select` click handler.

thangdinh replied:
I refactored it but right now it is not relevant (temporarily) because of the highlighted item bug, I will postpone the showing dropdown in multi-select.

azirbel commented on src/select.coffee:
```diff
@@ -346,11 +371,26 @@ Ember.AddeparMixins.ResizeHandlerMixin,
+    # [Spacebar, Enter, Up, Down, 'A'..'Z','a..z','0..9']
+    acceptedKeys = @get 'acceptedKeys'
+    if acceptedKeys[event.keyCode] and not @get 'showDropdown'
+      return @set('showDropdown', yes)
```
the result of a `@set` isn't very interesting; I think you should `set` and then `return` on a new line.

thangdinh replied:
Done

azirbel commented on tests/helpers/test_helpers.coffee:
```diff
@@ -117,6 +121,7 @@ Ember.Test.registerHelper 'mouseDown', _mouseDown
 Ember.Test.registerHelper 'isPresent', _isPresent
 Ember.Test.registerHelper 'isNotPresent', _isNotPresent
 Ember.Test.registerHelper 'isVisible', _isVisible
+Ember.Test.registerHelper 'isFocus', _isFocus
```
should rename to `isFocused`

thangdinh replied:
Done

azirbel commented on tests/unit/multi_select_test.coffee:
```diff
@@ -0,0 +1,96 @@
+  expect 17
+  selectedText = null
+
+  Ember.run ->
```
do we really need `Ember.run`? I don't understand enough about testing here, but I was able to remove it in an Iverson test I wrote recently.

thangdinh replied:
I don't know about it either. 
@cyril-sf @igillis do you have any idea?

thangdinh replied:
I removed it and it ran fine.

azirbel commented on tests/unit/multi_select_test.coffee:
```diff
@@ -0,0 +1,96 @@
+      # test pressing ENTER key to open dropdown
+      keyEvent(multiSelectComponent, 'keydown', 13)
+    .then ->
+      ok isVisible(find '.ember-select-results', multiSelectComponent, 'Dropdown list should appear after pressing Enter')
```
you can keep within the 80 character limit in many places just by putting a newline before the description for these checks. In our tests we usually put these on separate lines anyway.

azirbel replied:
check that it is initially not visible

thangdinh replied:
Done

azirbel commented on tests/unit/multi_select_test.coffee:
```diff
@@ -0,0 +1,96 @@
+      # test selecting option using ENTER key
+      selectedText = $(resultItems[0]).text().trim()
+
+      keyEvent(multiSelectComponent, 'keydown', 13)
```
you should make a helper for these `keyEvent`s. Something like `pressEnter(multiSelectComponent)` would read better here.

thangdinh replied:
Done

azirbel commented on tests/unit/multi_select_test.coffee:
```diff
@@ -0,0 +1,96 @@
+      # test if selected Item is actually selected
+      resultItems = find '.ember-select-search-choice', multiSelectComponent
+      currentText = $(resultItems[resultItems.length - 1]).text().trim()
+      # trim the special character "x" at the end of the select choice text
```
why is there a special character "x"?

thangdinh replied:
it's the close button next to selected options (pills)

azirbel replied:
ok, please add a comment to explain that :)

thangdinh replied:
Done

azirbel commented on tests/unit/select_test.coffee:
```diff
@@ -0,0 +1,120 @@
+      # test pressing ENTER key to open dropdown
+      keyEvent(selectComponent, 'keydown', 13)
+    .then ->
+      ok isVisible(find '.ember-select-results', selectComponent, 'Dropdown list should appear after pressing Enter')
```
check that it is initially not visible

thangdinh replied:
is the line 62 sufficient, @azirbel ?

azirbel replied:
oh, yeah! was that there the whole time? must have missed it

thangdinh replied:
Done

azirbel commented on tests/unit/select_test.coffee:
```diff
@@ -0,0 +1,120 @@
+
+      # test pressing UP arrowkey to navigate selection down
+      keyEvent(selectComponent, 'keydown', 38)
+
```
got some extra newlines in this file. I liked your style before where you'd do the `keyEvent` (or whatever action) and the `.then` on the following line.

thangdinh replied:
Done

azirbel commented on tests/unit/select_test.coffee:
```diff
@@ -0,0 +1,120 @@
+      keyEvent(selectComponent, 'keydown', 97)
+
+    .then ->
+      ok isVisible(find '.ember-select-results', selectComponent, 'Dropdown list should appear after pressing a letter')
```
check that it isn't there before you start typing

thangdinh replied:
Is it in line 90?

azirbel replied:
again, yes ... my mistake. looks good

azirbel commented:
This is great. I think my comments are fairly easy to address (but I hope you do address them; there are some old comments still there which I still think are valid). LGTMBC.

thangdinh commented:
@azirbel , which are old comments still there which are valid? I thought I addressed all of them. Have I missed any of them? Could you please help me point out which ones you found if you still remember?

azirbel commented:
Added a new reply to the "old valid" comment so you don't miss it.

thangdinh commented on src/multi_select.coffee:
```diff
@@ -22,6 +22,11 @@ Ember.Widgets.MultiSelectComponent = Ember.Widgets.SelectComponent.extend
+  # disable tabindex of the component container to set focus directly to
+  # the input field, which is always visible. This helps reducing one tab
+  # step to navigate back to the previous component
+  tabindex: -1
```
@azirbel I already had the tabindex here overridden as -1 as you mentioned in the other comment

azirbel replied:
oh, ok - great! Since the other comment was still around, I figured you hadn't addressed it; you might want to comment "done" or something as a reply next time so I can easily tell the comment has been resolved when I skim through.

thangdinh replied:
It's my bad. I'll do it from now on. Thanks.

cyril-sf commented on Gruntfile.coffee:
```diff
@@ -34,15 +34,15 @@ module.exports = (grunt) ->
         configFile: 'karma.conf.js'
         singleRun: false
-        exclude: ['build/src/ember_widgets.js', 'tests/functional/*.js', 'tests/integration/*.js'],
+        exclude: ['build/src/ember_widgets.js', 'build/tests/functional/*.js', 'build/tests/integration/*.js'],
```
If you rebase, that should be fixed on master

cyril-sf commented on src/mixins.coffee:
```diff
@@ -1,3 +1,16 @@
+Ember.Widgets.KeyboardHelper = Ember.Mixin.create
```
Much needed :)

cyril-sf commented on src/multi_select.coffee:
```diff
@@ -1,4 +1,4 @@
-Ember.Widgets.MultiSelectOptionView = Ember.View.extend
```
What are the chances that someone was relying on that class name?

thangdinh replied:
@cyril-sf You are right. But the original name is not very well-written. We have the SelectionOptionView in select component indicating the items in the dropdown while here it is the selected items (selected pills) which has a completely different meaning.
What should we do in this case?

cyril-sf commented on src/multi_select.coffee:
```diff
@@ -15,12 +15,24 @@ Ember.Widgets.MultiSelectOptionView = Ember.View.extend
+Ember.Widgets.MultiSelectOptionView = Ember.Widgets.SelectOptionView.extend
+  processDropDownShown: ->
+    @_super()
+    @get('controller').focusTextField()
```
A controller should not manipulate the DOM, it breaks the separation of concerns

cyril-sf replied:
It seems like `controller` is actually a component.

thangdinh replied:
Yes, it is a component.

cyril-sf commented on src/multi_select.coffee:
```diff
@@ -15,12 +15,24 @@ Ember.Widgets.MultiSelectOptionView = Ember.View.extend
   choicesFieldClass: ''
   placeholder: undefined
   persistentPlaceholder: undefined
+  resetQueryOnSelect: true
```
`yes` for our CS lover friends

thangdinh replied:
It is in master. I'll change it.

cyril-sf commented on src/select.coffee:
```diff
@@ -166,8 +182,7 @@ Ember.AddeparMixins.ResizeHandlerMixin,
     itemViewClass = @get 'itemViewClass'
-    if typeof itemViewClass is 'string'
-      return Ember.get itemViewClass
+    return Ember.get itemViewClass if typeof itemViewClass is 'string'
```
Not a huge fan of this. This assumes that the string represents the global path to the class. But you didn't write it, so we can revisit that later.

cyril-sf commented on src/select.coffee:
```diff
@@ -346,11 +367,27 @@ Ember.AddeparMixins.ResizeHandlerMixin,
   .property 'selectableOptions.[]', 'highlightedIndex'
 
+  setFocus: ->
+    activeElem = $(document.activeElement)[0]
```
Why do you need to go through jQuery?

thangdinh replied:
@cyril-sf I don't know the other ways. Would you please tell me some?

cyril-sf commented on tests/unit/multi_select_test.coffee:
```diff
@@ -0,0 +1,129 @@
+  textField = find '.ember-text-field', multiSelectComponent
+
+  validateDropdownVisible = (messageVisible) ->
+    ok isVisible(find '.ember-select-results', multiSelectComponent,
```
Shoudln't that be `ok isVisible(), messageVisible`?

thangdinh replied:
:+1:  it works and I didn't notice it. Thanks.

cyril-sf commented on tests/unit/multi_select_test.coffee:
```diff
@@ -0,0 +1,129 @@
+      selecting using Enter')
+    # test if selected Item is actually selected
+    resultItems = find '.ember-select-search-choice', multiSelectComponent
+    currentText = $(resultItems[resultItems.length - 1]).text().trim()
```
```cs
currentText = $(resultItems[resultItems.length - 1]).text().replace(/ x$/, '').trim()
equal(currentText, selectedText
```

thangdinh replied:
:+1:

cyril-sf commented on tests/unit/multi_select_test.coffee:
```diff
@@ -0,0 +1,129 @@
+      using mouse')
+
+    # test deleting one selected item using mouse
+    closeButtons = find '.ember-select-search-choice-close', multiSelectComponent
```
You should be able to simply write 

```cs
click  '.ember-select-search-choice-close', multiSelectComponent
```

thangdinh replied:
@cyril-sf which item will it click on assuming that `find '.ember-select-search-choice-close', multiSelectComponent` returns multiple results. I guess it will click on the 1st one?

cyril-sf commented on tests/unit/select_test.coffee:
```diff
@@ -0,0 +1,131 @@
+
+  validateDropdownHidden('Dropdown list should not exist at the beginning')
+  selectComponent.focus()
+  wait().then ->
```
Use `andThen` instead ( see http://guides.emberjs.com/v1.11.0/testing/test-helpers/#toc_wait-helpers )

`andThen` is the public API than can potentially do more in the future

cyril-sf replied:
Actually, `keyEvent` is an asynchronous helper, which means it can be chained without using `then`, `wait` or `andThen`

So we would probably better writing

```cs
append()
...
andThen() ->
    validateDropdownHidden
    selectComponent.focus()

pressEnter selectComponent
andThen ->
    validateDropdownVisible

pressDown selectComponent
andThen ->
    ....
```

cyril-sf commented:
The code looks good. I haven't tested like @igillis or @azirbel though.

There are a few comments remaining.

thangdinh commented on src/multi_select.coffee:
```diff
@@ -15,12 +15,24 @@ Ember.Widgets.MultiSelectOptionView = Ember.View.extend
+  # disable tabindex of the component container to set focus directly to
+  # the input field, which is always visible. This helps reducing one tab
+  # step to navigate back to the previous component
+  tabindex: -1
```
@cyril-sf it's addressed here.

cyril-sf replied:
:+1:

cyril-sf commented:
Is this the final working code?

If so, please rebase it the way you want it to be merged.

thangdinh commented:
@cyril-sf I squashed and rebased. Would you please help me review it?

thangdinh commented:
@cyril-sf when you have time, would you please help me review it? I need LGTM for QA to process :)

cyril-sf commented on src/select.coffee:
```diff
@@ -107,6 +110,19 @@ Ember.AddeparMixins.ResizeHandlerMixin,
+    keySet = _.union([@KEY_CODES.ENTER, @KEY_CODES.SPACEBAR],
+      [@KEY_CODES.DOWN, @KEY_CODES.UP],[65..90],[97..122],[48..57])
+    keySet.forEach (key) ->
+      mappedKeys[key] = true
```
`yes`

cyril-sf commented on tests/unit/multi_select_test.coffee:
```diff
@@ -32,3 +38,111 @@ test 'Query should be reset on select if resetQueryOnSelect is true', ->
+    resultItems = find '.ember-select-search-choice', multiSelectComponent
+    equal(resultItems.length, 1,
+      'There should be 1 selected item after deleting one')
+    validateTextFieldFocus('Dropdown list should still be shown after
```
```
   validateTextFieldFocus(
      'Dropdown list should still be shown after deleting using keyboard',
      'Text field should be focused after pressing delete')
```

is easier to read

thangdinh replied:
:+1:

cyril-sf commented on tests/unit/select_test.coffee:
```diff
@@ -38,6 +44,78 @@ test 'Test sorted filter', ->
+  andThen ->
+    validateDropdownVisible('Dropdown list should appear after pressing Enter')
+
+    # test pressing DOWN arrowkey to navigate selection down
```
extra space not needed

cyril-sf commented:
Some minor comments, but LGTM!

Nice work!

thangdinh commented:
Just fixed the search matcher, and added tests. Could you please help me review it, @igillis?

igillis commented:
lgtmbc

thangdinh commented:
Merged
